### PR TITLE
TST: sparse: Prepare `test_base.py` tests to work for sparray or spmatrix classes

### PR DESCRIPTION
**Pull request by dschult**
This is the 2nd and perhaps the ugliest of the 3 PRs to update `test_base.py` for testing either sparray or spmatrix classes. It sets up the tests to check `isinstance(spcreator, sparray)` and do the right test.

The 1st PR of the three, #20889 focuses on tests of the matmul and mul operators.  The 3rd makes the changes from `*_matrix` to `*_object` and `asmatrix` to `asobject` so that it's easy to define `*_object` and `asobject` for easy switching. We don't see test_base being used to test both sparray and spmatrix at once except in a test designed just for that.

I apologize for the difficulty and importance of this kind of long hard-to-read PR. That's why I have split it up. 

This PR does the following:
- adds a property to the test classes: `is_array_test` to ease checking whether we are testing sparray or spmatrix
- fix tests for str and repr to test sparray or spmatrix
- bypass tests of in-place reshape for sparray and check exception instead
- switch tests of `getrow, getcol, asfptype` to  test `_getrow, _getcol, _asfptype` and  use `nnz` or `_getnnz` if testing `getnnz`
- change `test_sum()` test to switch `matrix` to `asmatrix` or `np.array` [matrix is primarily used in these tests to ensure a 2d array result]
- change `keepdims` args to depend on spcreator type
- change `test_pow` to account for elementwise vs matrix power
- replace `matrix` in `test_matvec` where it is only needed for 2d shape and change to `asmatrix` where possible (for easy shift to `asobject`)
- remove `matrix.A` usage near line 2738 because `assert_array_equal` doesn't need it and `.A` is removed for sparray
- one very minor change to `_data.py` for `__imul__` to align the `imul` methods' code across formats and make more readable

I will keep this PR as a draft until #20889 is merged, and will update it as changes to that PR are made. This PR is built on that one.

### CODE REVIEW

**Comment by dschult** [comment]
I have rebased on main after the merge of #20889
The OP list of changes in this PR still holds.

The PR is no longer "draft" and is ready for review!

**Review by perimosocordiae — CHANGES_REQUESTED** [review]
Everything is good here, except for testing `_getrow()`, `_getcol()`, and `_getnnz()` functions directly. I'd rather we tested the public-facing API only, which means skipping the test for the `is_array_test == True` case.

**Comment by dschult** [comment]
I have reverted the changes that test `_getrow` and similar so we still test `getrow` with spmatrices, and will make the changes in #20895 test them only for spmatrix.

I have also added a number of changes to make the test_base.py tests pass for sparrays. The process of testing sparrays found a couple of bugs:
- For `A[3,:] = B[2,:]` the results of indexing B is 1-D so we're setting some of the values in A with a sparse 1D object. That revealed some code in `_index.py` that still assumed a 2D shape. 
- `_validate_indices` wasn't handling the case of `A[..., 1:3]` with a ellipsis and a slice. It was adding the wrong number in the shape of the result. 
Luckily both these bugs were in `_index.py` and not yet released. So it's important to get direct testing of sparray included.

Most of the changes are in the tests because sparrays don't return the same thing as spmatrices. I encourage you to look at the changes one commit at a time to reduce overload in review.

The goal of this PR is to get the tests themselves ready to start testing sparrays. But it doesn't actually start testing them until the following PR #20895. That PR has been updated from having a switch for sparray or spmatrix. It now actually creates two classes for each format and runs the tests on **both** sparray and spmatrix. It doesn't do this for `Test64Bit` because that would make the tests take too long. But those changes are separated from these to ease review.

File: scipy/sparse/tests/test_base.py [hunk 1]
```diff
@@ -5291,32 +5337,31 @@ def test_downcast_intp(self):
...
+            assert_raises(AssertionError, a.count_nonzero, axis=1)
             assert_raises(AssertionError, a.sum, axis=0)
 
             a = csr_matrix([[1, 2, 3], [3, 4, 6]])
-            assert_raises(AssertionError, a.getnnz, axis=0)
+            assert_raises(AssertionError, a.count_nonzero, axis=0)
 
             a = coo_matrix([[1, 2, 3], [3, 4, 5]])
-            assert_raises(AssertionError, a.getnnz, axis=0)
+            assert_raises(AssertionError, a.count_nonzero, axis=0)
```
Comment (perimosocordiae) [inline scipy/sparse/tests/test_base.py:5351]: To keep the same coverage, we should have:

```
CSR.sum(axis=1)
CSC.sum(axis=0)
COO.sum_duplicates()
```

**Review by perimosocordiae — APPROVED** [review]

File: scipy/sparse/tests/test_base.py (hunk 1 above)
Comment (dschult) [reply scipy/sparse/tests/test_base.py]: *Context for those not at the sparse array meeting:
A few lines above this suggestion is a comment stating that the purpose of these tests is to check that the np.bincount and np.reduceat code deals with intp downcasts so those functions work on 32-bit systems when using 64-bit indices.*

It looks like testing `COO.sum_duplicates()` actually **expands** the coverage. We didn't used to test the `reduceat` that is used in COO.sum_duplicates. And including it created errors. The errors are fixed by using `downcast_intp_index` in COO.sum_duplicates. So I have added those tests and fixed the errors that it found! Thanks for this pointer!

Secondly, I think we weren't seeing the main branch when we were searching for `bincount` at the meeting today. On the main branch, the function `count_nonzero` uses bincount (since June 4 -- PR 20857 and it will be released in v1.15). The new `axis` parameter in `count_nonzero` mimics the code in `_getnnz` quite closely. Replacing the `getnnz` test with `count_nonzero` test is thus both needed and desired since the code is the same in terms of testing index downcasting.

So, my commit to address this suggestion: 
- **bincount**: keeps the `count_nonzero` tests for CSC, CSR and COO as added in this PR
- **reduceat**: keeps the tests of `CSC.sum(axis=0)`, adds the test of `CSR.sum(axis=1)` (probably left out because it is the same code).
- **reduceat**: adds the `COO.sum_duplicates` tests for COO. To make this pass, we must fix the downcasting problem in `COO.sum_duplicates` by adding `downcast_intp_index`.

I also noticed that the `sum_duplicates` test will require that `has_canonical_format` be set to False to make the test actually call `reduceat` so I added that.

In summary we will test:
```python
CSR.sum(axis=1)
CSR.count_nonzero(axis=0)
CSC.sum(axis=0)
CSC.count_nonzero(axis=1)
has_canonical_format = False
COO.sum_duplicates()
COO.count_nonzero(axis=0)   # no need to test both axis=0 and axis=1 because same situation in both
```

**Comment by dschult** [comment]
The follow-up PR (https://github.com/scipy/scipy/pull/20895) to this one has also been approved by @perimosocordiae and is rebased on top of this one so includes all these changes. So, I'm going to close this PR -- it can be used for folks who want to review that PR in stages rather than all at once. So I'll leave it open for another day or so for that purpose.

**Comment by dschult** [comment]
See #20895
